app/src/main/java/com/tolkiana/tacoapp/TacoApplication.kt

```diff
@@ -0,0 +1,22 @@
+package com.tolkiana.tacoapp
+
+import android.app.Application
+import com.android.volley.toolbox.ImageLoader
+import com.android.volley.toolbox.Volley
+import com.tolkiana.tacoapp.utilities.BitmapCache
+import com.tolkiana.tacoapp.utilities.DataParser
+
+/**
+ * Created by tolkiana on 9/10/17.
+ */
+class TacoApplication: Application() {
+
+    var imageLoader: ImageLoader? = null
+    var dataParser: DataParser? = null
+
+    override fun onCreate() {
+        super.onCreate()
+        imageLoader = ImageLoader(Volley.newRequestQueue(this), BitmapCache())
```

Comment:
This one is a little trickier, because I think you're after an `ImageLoader` Singleton, but Kotlin Singleton `object` classes cannot take any constructor parameters, so you wouldn't be able to initialize the Volley request queue in the constructor (since Volley requires a `context`).

One approach is to create a more conventional Singleton (something that would be more like the Java approach and modified from https://stackoverflow.com/a/45943282/1798762):

```kt
// I just picked a name, not sure what to name this exactly :)
class ApplicationImageLoader {
    companion object {

        @Volatile private var INSTANCE: ImageLoader? = null

        fun getInstance(context: Context): ImageLoader =
                INSTANCE ?: synchronized(this) {
                    INSTANCE ?: buildImageLoader(context.applicationContext).also { INSTANCE = it }
                }

        private fun buildImageLoader(context: Context) =
                ImageLoader(Volley.newRequestQueue(context), BitmapCache())
    }
}
```

When `ApplicationImageLoader.getInstance(context)` is first called, the class will create an `ImageLoader` instance. All future calls will just return the instance. Using this Singleton looks like:

```kt
val productAdapter = ProductAdapter(tacoList, ApplicationImageLoader.getInstance(this))
```

Reply:
Thank you for the comments! I like this approach, this also answers the question I had which is, what is the best practice when it comes to use the Application class to hold instances for objects that otherwise could be expensive to create multiple times. 

I have one particular question in this example, what `Volatile` means?